### PR TITLE
FunctionRef

### PR DESCRIPTION
**Pull request by Horusiath**
This is port of https://github.com/akka/akka/pull/19487

It allows to use Akka.Stream stage actor refs to be location transparent. It's necessary step before #3321

### CODE REVIEW

**Comment by Horusiath** [comment]
I see that some of the @Aaronontheweb changes from rebased commits have leaked during rebase. I'll fix them later.

File: src/core/Akka.Streams/Stage/GraphStage.cs [hunk 1]
```diff
@@ -1619,21 +1619,35 @@ protected Action GetAsyncCallback(Action handler)
...
 
-            return _stageActorRef;
+            return _stageActor;
         }
 
+        /// <summary>
+        /// Override and return a name to be given to the StageActor of this stage.
+        /// 
+        /// This method will be only invoked and used once, during the first <see cref="GetStageActor"/>
+        /// invocation whichc reates the actor, since subsequent `getStageActors` calls function
```
Comment (marcpiechura) [inline src/core/Akka.Streams/Stage/GraphStage.cs]: "which creates"

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt [hunk 2]
```diff
@@ -4400,21 +4402,17 @@ namespace Akka.Streams.Stage
         protected PushStage() { }
         public virtual Akka.Streams.Stage.ISyncDirective OnPull(Akka.Streams.Stage.IContext<TOut> context) { }
     }
-    public sealed class StageActorRef : Akka.Actor.MinimalActorRef
+    public sealed class StageActor
```
Comment (marcpiechura) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt]: Should we move this to 1.4 for the breaking changes?

File: src/core/Akka.Tests/Actor/FunctionRefSpec.cs [hunk 3]
```diff
@@ -0,0 +1,253 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="FunctionRefSpec.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2018 Akka.NET project <https://github.com/akkadotnet/akka.net>
```
Comment (marcpiechura) [inline src/core/Akka.Tests/Actor/FunctionRefSpec.cs]: needs the year range for Akka.NET too.

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt (hunk 2 above)
Comment (Horusiath) [reply src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt]: Depends on how do we threat a `[ApiMayChange]` attribute - only methods with that API have been changed. We can put it to v1.4 - from my perspective it's necessary for purposes of StreamRefs, which probably will land somewhere in v1.4 as well.

File: src/core/Akka.Tests/Actor/FunctionRefSpec.cs (hunk 3 above)
Comment (Aaronontheweb) [reply src/core/Akka.Tests/Actor/FunctionRefSpec.cs]: @marcpiechura going to be submitting an update to all of the headers soon

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt (hunk 2 above)
Comment (marcpiechura) [reply src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt]: Except for the `StageActorRef` property itself

**Comment by Horusiath** [comment]
Ok @akkadotnet/core  , do we want to merge those changes? I know that we have some breaking changes in area around StageActorRefs, so my question is should we merge it to dev or v1.4 branch?

**Pull request merged**